lib/chain.js

```diff
@@ -2,80 +2,90 @@
 
 module.exports = (api) => {
 
-  const resolver = (resolve, reject) => (
-    (err, result) => {
-      if (err) reject(err);
-      else resolve(result);
-    }
-  );
-
-  class ArrayChain {
-
-    constructor(array) {
-      this._promise = Promise.resolve(array);
-    }
-
-    then(fn) {
-      return this._promise.then(fn);
+  const async = op => {
+    switch (op) {
+      case 'map': return api.metasync.map;
+      case 'filter': return api.metasync.filter;
+      case 'reduce': return api.metasync.reduce;
+      case 'each': return api.metasync.each;
+      case 'series': return api.metasync.series;
+      case 'find': return api.metasync.find;
     }
+  };
 
-    catch(fn) {
-      return this._promise.catch(fn);
-    }
-
-    map(fn) {
-      this._update((array, resolve, reject) => {
-        api.metasync.map(array, fn, resolver(resolve, reject));
-      });
-      return this;
-    }
-
-    filter(fn) {
-      this._update((array, resolve, reject) => {
-        api.metasync.filter(array, fn, resolver(resolve, reject));
-      });
-      return this;
-    }
-
-    reduce(fn, initial) {
-      this._update((array, resolve, reject) => {
-        api.metasync.reduce(array, fn, resolver(resolve, reject), initial);
-      });
-      return this._promise;
-    }
-
-    each(fn) {
-      this._update((array, resolve, reject) => {
-        api.metasync.each(array, fn, resolver(resolve, reject));
-      });
-      return this._promise;
-    }
-
-    series(fn) {
-      this._update((array, resolve, reject) => {
-        api.metasync.series(array, fn, resolver(resolve, reject));
-      });
-      return this._promise;
-    }
+  function ArrayChain(array) {
+    this.array = array;
+    this.chain = [];
```

Comment:
Even theoretically array gives more performance, but in V8 arrays are highly optimized, so better you will test it and I expect arrays to be best.